### PR TITLE
Lp-2816  Implementation for exporting emails

### PR DESCRIPTION
**Pull request by javeriaejaz-arbisoft**
[LP-2816](https://philanthropyu.atlassian.net/browse/LP-2816)

[Theme PR](https://github.com/OmnipreneurshipAcademy/adg-edx-theme/pull/477)

***

#### Description:

- On clicking Export button, a CSV file is downloaded with emails of the accepted applicants. 

- User can select following different options (All Emails, Last 30 Days, Last 6 Months, Last 1 Year, Custom)

### CODE REVIEW

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2795/workflows/workflows_id/jobs/2795/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/applications/constants.py (100%)
openedx/adg/admin_dashboard/applications/helpers.py (100%)
openedx/adg/admin_dashboard/applications/tests/test_helpers.py (100%)
openedx/adg/admin_dashboard/applications/tests/test_views.py (100%)
openedx/adg/admin_dashboard/applications/views.py (75.0%): Missing lines 121-123,125
-------------
Total:   69 lines
Missing: 4 lines
Coverage: 94%
-------------
```

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2796/workflows/workflows_id/jobs/2796/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/applications/constants.py (100%)
openedx/adg/admin_dashboard/applications/helpers.py (100%)
openedx/adg/admin_dashboard/applications/tests/test_helpers.py (100%)
openedx/adg/admin_dashboard/applications/tests/test_views.py (100%)
openedx/adg/admin_dashboard/applications/views.py (75.0%): Missing lines 121-123,125
-------------
Total:   69 lines
Missing: 4 lines
Coverage: 94%
-------------
```

**Review by annas-usman — APPROVED** [review]
LGTM, just added a small suggestion

File: openedx/adg/admin_dashboard/applications/helpers.py [hunk 1]
```diff
@@ -224,3 +233,52 @@ def get_application_review_url(application_id, query_params_string):
...
+    """
+    Convert day/month/year to a date object.
+
+    Arguments:
+        day (string): day of the date
+        month (string): month of the date
+        year (string): year of the date
+
+    Returns:
+        _date: date object if it is valid, otherwise None
```
Comment (aightmunam) [inline openedx/adg/admin_dashboard/applications/helpers.py:248]: _date is not a type, let's call it `datetime.date` or `date`

File: openedx/adg/admin_dashboard/applications/helpers.py [hunk 2]
```diff
@@ -224,3 +233,52 @@ def get_application_review_url(application_id, query_params_string):
...
+        to_day (date): Ending date
+    """
+    to_day = now()
+
+    if duration == LAST_6_MONTHS:
+        from_day = to_day - relativedelta(months=6)
+    elif duration == LAST_30_DAYS:
+        from_day = to_day - relativedelta(days=30)
+    elif duration == LAST_1_YEAR:
+        from_day = to_day - relativedelta(years=1)
```
Comment (aightmunam) [inline openedx/adg/admin_dashboard/applications/helpers.py:277]: Can we achieve all of this by using a dict?

File: openedx/adg/admin_dashboard/applications/views.py [hunk 3]
```diff
@@ -101,6 +104,36 @@ def setup(self, request, *args, **kwargs):
...
         self.application_type = kwargs.get('application_type')
 
+    def get(self, request, *args, **kwargs):
+        """
+        Attaches csv report for download with response.
+
+        Returns:
+            HttpResponse: Response with grades report.
+        """
+        if not request.GET.get('download_report') or request.GET.get('download_report') == 'False':
```
Comment (aightmunam) [inline openedx/adg/admin_dashboard/applications/views.py:114]: We do not need to get the value twice from the request_dict.
Perhaps something like `request.GET.get('download_report', False) == False` will work for us here.

File: openedx/adg/admin_dashboard/applications/views.py [hunk 4]
```diff
@@ -101,6 +104,36 @@ def setup(self, request, *args, **kwargs):
...
+            return super().get(request, *args, **kwargs)
+
+        applications = self.get_queryset()
+        export_email_range = request.GET.get('export_email_range')
+
+        if export_email_range != 'all':
```
Comment (aightmunam) [inline openedx/adg/admin_dashboard/applications/views.py:120]: Let's use a constant to describe 'all' and use that here.

File: openedx/adg/admin_dashboard/applications/views.py [hunk 5]
```diff
@@ -101,6 +104,36 @@ def setup(self, request, *args, **kwargs):
...
+
+            applications = applications.filter(
+                user__application_hub__submission_date__gte=from_day, user__application_hub__submission_date__lte=to_day
+            )
+
+        emails = list(applications.values_list('user__email'))
+        emails.insert(0, ('Emails',))
+
+        response = get_http_response_with_csv_file_attachment(
+            emails, file_name=f'accepted_applicants_{now().date()}.csv'
```
Comment (aightmunam) [inline openedx/adg/admin_dashboard/applications/views.py:133]: Let's create a constant for filename naming and then import and format that constant here with the current date.

Comment (aightmunam) [inline openedx/adg/admin_dashboard/applications/views.py:133]: Let's create a constant for filename naming and then import and format that constant here with the current date.

**Review by aightmunam — CHANGES_REQUESTED** [review]

File: openedx/adg/admin_dashboard/applications/helpers.py (hunk 2 above)
Comment (javeriaejaz-arbisoft) [reply openedx/adg/admin_dashboard/applications/helpers.py:277]: It won't be consistent if we use dict for 3 options and if/else for the fourth

Comment (aightmunam) [reply openedx/adg/admin_dashboard/applications/helpers.py:277]: imo an `if-else` with a dict is still a cleaner approach. But it's completely your choice!

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2798/workflows/workflows_id/jobs/2798/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/applications/constants.py (100%)
openedx/adg/admin_dashboard/applications/helpers.py (100%)
openedx/adg/admin_dashboard/applications/tests/test_helpers.py (100%)
openedx/adg/admin_dashboard/applications/tests/test_views.py (100%)
openedx/adg/admin_dashboard/applications/views.py (73.3%): Missing lines 122-124,126
-------------
Total:   67 lines
Missing: 4 lines
Coverage: 94%
-------------
```

File: openedx/adg/admin_dashboard/applications/views.py [hunk 6]
```diff
@@ -101,6 +105,34 @@ def setup(self, request, *args, **kwargs):
...
         self.application_type = kwargs.get('application_type')
 
+    def get(self, request, *args, **kwargs):
+        """
+        Attaches csv report for download with response.
+
+        Returns:
+            HttpResponse: Response with grades report.
+        """
+        if request.GET.get('download_report', 'False') == 'False':
```
Comment (aightmunam) [inline openedx/adg/admin_dashboard/applications/views.py:115]: Can we kindly use False and True instead of 'False' and 'True', if possible?

**Review by aightmunam — CHANGES_REQUESTED** [review]

File: openedx/adg/admin_dashboard/applications/helpers.py (hunk 1 above)
Comment (aightmunam) [reply openedx/adg/admin_dashboard/applications/helpers.py]: @javeriaejaz-arbisoft this still needs to be updated according to [google format docstring](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2800/workflows/workflows_id/jobs/2800/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/applications/constants.py (100%)
openedx/adg/admin_dashboard/applications/helpers.py (100%)
openedx/adg/admin_dashboard/applications/tests/test_helpers.py (100%)
openedx/adg/admin_dashboard/applications/tests/test_views.py (100%)
openedx/adg/admin_dashboard/applications/views.py (75.0%): Missing lines 123-125,127
-------------
Total:   67 lines
Missing: 4 lines
Coverage: 94%
-------------
```

**Comment by omnipreneurshipA** [comment]
## Diff Coverage Report - ADG
Check-out complete **[report](https://app.circleci.com/pipelines/github/OmnipreneurshipAcademy/edx-platform/2801/workflows/workflows_id/jobs/2801/artifacts)**. Coverage !!! :rocket:
```
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
openedx/adg/admin_dashboard/applications/constants.py (100%)
openedx/adg/admin_dashboard/applications/helpers.py (100%)
openedx/adg/admin_dashboard/applications/tests/test_helpers.py (100%)
openedx/adg/admin_dashboard/applications/tests/test_views.py (100%)
openedx/adg/admin_dashboard/applications/views.py (75.0%): Missing lines 123-125,127
-------------
Total:   67 lines
Missing: 4 lines
Coverage: 94%
-------------
```

**Pull request merged**